quadruped_pympc/controllers/Controllers.py

```diff
@@ -0,0 +1,543 @@
+    def get_forward_action(self,simulation=True,env=None):
+        if simulation == True:
+            nmpc_GRFs = self.nmpc_GRFs
+        else:
+            ##suff
+            breakpoint()
+        
+        # Compute Stance Torque ---------------------------------------------------------------------------
+        feet_jac = self.env.feet_jacobians(frame='world', return_rot_jac=False)
+        # Compute feet velocities
+        feet_vel = LegsAttr(**{leg_name: feet_jac[leg_name] @ self.env.mjData.qvel for leg_name in self.legs_order})
+        # Compute jacobian derivatives of the contact points
+        jac_feet_dot = (feet_jac - self.jac_feet_prev) / self.simulation_dt  # Finite difference approximation
+        self.jac_feet_prev = feet_jac  # Update previous Jacobians
+        # Compute the torque with the contact jacobian (-J.T @ f)   J: R^nv -> R^3,   f: R^3
+        self.tau.FL = -np.matmul(feet_jac.FL[:, self.env.legs_qvel_idx.FL].T, nmpc_GRFs.FL)
+        self.tau.FR = -np.matmul(feet_jac.FR[:, self.env.legs_qvel_idx.FR].T, nmpc_GRFs.FR)
+        self.tau.RL = -np.matmul(feet_jac.RL[:, self.env.legs_qvel_idx.RL].T, nmpc_GRFs.RL)
+        self.tau.RR = -np.matmul(feet_jac.RR[:, self.env.legs_qvel_idx.RR].T, nmpc_GRFs.RR)
+
+        # ---------------------------------------------------------------------------------------------------
+
+        # Compute Swing Torque ------------------------------------------------------------------------------
+        # TODO: Move contact sequence to labels FL, FR, RL, RR instead of a fixed indexing.
+        # The swing controller is in the end-effector space. For its computation,
+        # we save for simplicity joints position and velocities
+        qpos, qvel = self.env.mjData.qpos, self.env.mjData.qvel
+        # centrifugal, coriolis, gravity
+        legs_mass_matrix = self.env.legs_mass_matrix
+        legs_qfrc_bias = self.env.legs_qfrc_bias
+        self.stc.update_swing_time(self.current_contact, self.legs_order, self.simulation_dt)
+
+        for leg_id, leg_name in enumerate(self.legs_order):
+            if self.current_contact[leg_id] == 0:  # If in swing phase, compute the swing trajectory tracking control.
+                self.tau[leg_name], _, _ = self.stc.compute_swing_control(
+                    leg_id=leg_id,
+                    q_dot=qvel[self.env.legs_qvel_idx[leg_name]],
+                    J=feet_jac[leg_name][:, self.env.legs_qvel_idx[leg_name]],
+                    J_dot=jac_feet_dot[leg_name][:, self.env.legs_qvel_idx[leg_name]],
+                    lift_off=self.frg.lift_off_positions[leg_name],
+                    touch_down=self.nmpc_footholds[leg_name],
+                    foot_pos=self.feet_pos[leg_name],
+                    foot_vel=feet_vel[leg_name],
+                    h=legs_qfrc_bias[leg_name],
+                    mass_matrix=legs_mass_matrix[leg_name]
+                    )
+        tau=self.tau
+        return tau    
+
+    def reset(self,env):
```

Comment:
The `QuadrupedEnv` instance should never be passed to this class. Since if we use this class to control the real robot, we should not depend nor use the env instance. 

What each of these functions should have as signatures is the expected observations from the env. I.e, the feet positions, current contact etc

Reply:
Remove the use of env inside the code and set the correct signature for each controller function input